src/object/cli_ec.c

```diff
@@ -1339,6 +1353,155 @@ obj_ec_req_reasb(daos_obj_rw_t *args, daos_obj_id_t oid,
 	return rc;
 }
 
+static struct obj_ec_recov *
+obj_ec_recov_alloc(struct daos_oclass_attr *oca)
+{
+	struct obj_ec_recov	*recov;
+	unsigned short		 k = obj_ec_data_tgt_nr(oca);
+	unsigned short		 p = obj_ec_parity_tgt_nr(oca);
+	void			*buf, *tmp_ptr;
+	size_t			 struct_size, tbl_size, matrix_size;
+	size_t			 idx_size, list_size, err_size;
+
+	struct_size = roundup(sizeof(struct obj_ec_recov), 8);
+	tbl_size = k * p * 32;
+	matrix_size = roundup((k + p) * k, 8);
+	idx_size = roundup(sizeof(uint32_t) * k, 8);
+	list_size = roundup(sizeof(uint32_t) * p, 8);
+	err_size = roundup(sizeof(bool) * (k + p), 8);
+
+	D_ALLOC(buf, struct_size + tbl_size + 3 * matrix_size + idx_size +
+		     list_size + err_size);
+	if (buf == NULL)
+		return NULL;
+
+	tmp_ptr = buf;
+	recov = buf;
+	tmp_ptr += struct_size;
+	recov->er_gftbls = tmp_ptr;
+	tmp_ptr += tbl_size;
+	recov->er_de_matrix = tmp_ptr;
+	tmp_ptr += matrix_size;
+	recov->er_inv_matrix = tmp_ptr;
+	tmp_ptr += matrix_size;
+	recov->er_b_matrix = tmp_ptr;
+	tmp_ptr += matrix_size;
+	recov->er_dec_idx = tmp_ptr;
+	tmp_ptr += idx_size;
+	recov->er_err_list = tmp_ptr;
+	tmp_ptr += list_size;
+	recov->er_in_err = tmp_ptr;
+
+	return recov;
+}
+
+void
+obj_ec_recov_free(struct obj_reasb_req *reasb_req)
+{
+	if (reasb_req->orr_recov)
+		D_FREE(reasb_req->orr_recov);
+}
+
+static bool
+obj_ec_err_match(uint32_t nerrs, uint32_t *err_list1, uint32_t *err_list2)
+{
+	uint32_t	i;
+
+	for (i = 0; i < nerrs; i++) {
+		if (err_list1[i] != err_list2[i])
+			return false;
+	}
+	return true;
+}
+
+int
+obj_ec_recov_prep(struct obj_reasb_req *reasb_req, daos_obj_id_t oid,
```

Comment:
would be nice if you can add a description for this function.

Reply:
sure, I'll add some comment in following patches.